### PR TITLE
Add HeatmapLayer Component

### PR DESCRIPTION
**Pull request by zolrath**
Allows a HeatmapLayer to be created. Takes an array of
`google.maps.LatLng` objects to the prop `data`.

I'm not sure what events this should really be listening for so if you have any suggestions or things I need to change please let me know!

### CODE REVIEW

**Comment by cristiandley** [comment]
This would be awesome to be implemented. I was working on my own implementation and i found you made a pull request with the same idea. Thanks @zolrath i wish it merge soon.

**Comment by cristiandley** [comment]
@zolrath Matt, why are you not using /lib/ for this addon ?

**Comment by cristiandley** [comment]
@zolrath i made it work there are some issues, please compile to es5 the work you'v done

mdd

**Comment by cristiandley** [comment]
@zolrath i put your heatmaplayer on my own repo [https://www.npmjs.com/package/react-gl-maps](https://www.npmjs.com/package/react-gl-maps) just check the /lib/ its missing the babel compilation

**Comment by tomchentw** [comment]
@cristiandley for PRs, it should be clean enough for anyone to see the changes. `lib` are automatically generated by the release script during publish process.

File: src/addons/HeatmapLayer.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    // Event [onEventName]
+    ...heatmapLayerEventPropTypes,
+  }
+
+  // Public APIs
+  //
+  // https://developers.google.com/maps/documentation/javascript/3.exp/reference#HeatmapLayer
+  //
+  // [].map.call($0.querySelectorAll("tr>td>code"), function(it){ return it.textContent; }).filter(function(it){ return it.match(/^get/) && !it.match(/^getMap/); })
+  getData() { return this.state.directionsRenderer.getDirections(); }
```
Comment (tomchentw) [inline src/addons/HeatmapLayer.js]: Should this be removed?

**Comment by tomchentw** [comment]
@zolrath Just left one code review comment. Also, I'd like this component to be under `visualization` folder instead of `addons`. `addons` should be used only for 3rd party components.

Other than that, great work!

**Comment by zolrath** [comment]
I will make the requested changes right now, thanks!

**Comment by cristiandley** [comment]
@zolrath 

Can you add also the ` radius ` and  ` opacity ` properties ?

Thanks and i make my final review

**Comment by zolrath** [comment]
In the ControlledPropTypes function of the other components I see:

> // NOTICE!!!!!!
//
// Only expose those with getters & setters in the table as controlled props.
//
// https://developers.google.com/maps/documentation/javascript/3.exp/reference#HeatmapLayer

From what I can tell that means I should not expose `radius` and `opacity` as they are set in the Options object, not via a getter/setter in the Heatmap?

**Comment by tomchentw** [comment]
@cristiandley let's hold this on for `5.0.0` first?

**Comment by cristiandley** [comment]
@tomchentw yes! :)

**Comment by 11th** [comment]
Any progress on this?

**Comment by zolrath** [comment]
What changes need to be done to get this into the new version of react-google-maps?

**Comment by tomchentw** [comment]
Released [v5.1.0](https://github.com/tomchentw/react-google-maps/releases/tag/v5.1.0) on npm `next` tag.

**Comment by farhan687** [comment]
Hey @tomchentw 

How can I use heatmap layer? 
Npm package still shows version: 4.11.0. Any plan to release new version of npm?

**Comment by farhan687** [comment]
Hello everyone,
 I am using @beta tag, and I want to use heatmaplayer.  I have already tried by wrapping it inside `withGoogleMap` and also without it, but it's really strange that console shows following error `Uncaught TypeError: google.maps.HeatmapLayer is not a constructor`.

Here is my implementation:

```javascript
import { withGoogleMap, GoogleMap, Marker, HeatmapLayer } from "react-google-maps";

<HeatmapLayer
    data = {[
      37.782, -122.447,
     37.782, -122.445
    ]}
  />
```

Can anyone tell me if I am missing anything?

**Comment by matthill1** [comment]
Seconded farhan here, if anybody has successfully implemented the heatmap layer, would you mind posting an example of syntax? The documentation is scarce. Thank you.

**Comment by ribeiroguilherme** [comment]
+1 looking for the syntax example as well

**Comment by ribeiroguilherme** [comment]
@farhan687  did you figure out how to use? I'm with the same problem here

**Comment by ribeiroguilherme** [comment]
Hi guys.
I figured out the problem. In the HeatmapLayer file, is being created the heatmap using:
`new google.maps.HeatmapLayer` , but is incorrect this path.
Should be
`new google.maps.visualization.HeatmapLayer`

**Comment by ZephD** [comment]
Even with this change, I get "b.lat() is not a function".
Internally, the visualization library seems to _require_ a google.maps.LatLng, a simple {lat: 0, lng: 0} object does not suffice.
Since google is loaded async, creating new latlngs is very hard. Maybe the HeatmapLayer should convert the data to a LatLng object?

**Comment by ribeiroguilherme** [comment]
ZephD, you can create an array like
`array.push(new google.maps.LatLng(lat, long));` and pass it to 'data' property in HeatmapLayer.

If you load the google maps script in your index html file, should work fine. Here is working fine.. But we need to fix this bug. When we deploy to production, the node will download the lib with the wrong path.

**Comment by ZephD** [comment]
But I do not have google maps in the index file, because that'll be too easy. Google maps is loaded Async, so I do not have access to new google.maps.LatLng, hence the problem.

**Comment by ZephD** [comment]
#409

**Comment by GustavoMantuan** [comment]
Hello, I am using this way:

`
import HeatmapLayer from "react-google-maps/lib/visualization/HeatmapLayer";


const MapaPerfilW = withGoogleMap(props => (
    <GoogleMap
        defaultZoom={15}
        center={props.defaultCenter}
    >
        <HeatmapLayer
            data={[
                new google.maps.LatLng(37.782551, -122.445368),
                new google.maps.LatLng(37.782745, -122.444586),
                new google.maps.LatLng(37.782842, -122.443688),
                new google.maps.LatLng(37.782919, -122.442815),
                new google.maps.LatLng(37.782992, -122.442112),
                new google.maps.LatLng(37.783100, -122.441461)
            ]}
        />
    </GoogleMap>
));`

Version 7.1.0 working great.

**Comment by malindap92** [comment]
@GustavoMantuan could you tell me how you were able to access google.maps.LatLng? it gives me 'google' is undefined. Is there any library to import?

**Comment by MtPython** [comment]
@malindap92   add `/*global google*/` at the top of your **.js** file, and the next thing you know - use `<script src="https://maps.googleapis.com/maps/api/js?v=3.exp&key=**YOUR_KEY**&libraries=visualization"></script>` to get the other questions out of the way

**Comment by rvntone** [comment]
@GustavoMantuan I had this:

> /*global google*/
> import React, { Component } from 'react';
> 
> import { withGoogleMap, GoogleMap } from 'react-google-maps';
> import HeatmapLayer from 'react-google-maps/lib/visualization/HeatmapLayer';
> 
> import data from './data.json';
> 
> const HeatMap = withGoogleMap(props => {
>   return (
>     <GoogleMap defaultZoom={15} center={props.defaultCenter}>
>       <HeatmapLayer
>         data={[
>           new google.maps.LatLng(37.782551, -122.445368),
>           new google.maps.LatLng(37.782745, -122.444586),
>           new google.maps.LatLng(37.782842, -122.443688),
>           new google.maps.LatLng(37.782919, -122.442815),
>           new google.maps.LatLng(37.782992, -122.442112),
>           new google.maps.LatLng(37.7831, -122.441461),
>         ]}
>       />
>     </GoogleMap>
>   );
> });
> 
> class GeographicalInfo extends Component {
>   render() {
>     return (
>       <div>
>         Geographical Info
>         <HeatMap
>           containerElement={<div style={{ height: `100%` }} />}
>           mapElement={<div style={{ height: `100%` }} />}
>           defaultCenter={{ lat: 0, lng: 0 }}
>         />
>       </div>
>     );
>   }
> }
> export default GeographicalInfo;

And I got:

Uncaught (in promise) TypeError: Cannot read property 'HeatmapLayer' of undefined, it happened in the google library (google.maps.visualization is undefined).

> var heatmapLayer = new google.maps.visualization.HeatmapLayer((0, _extends3.default)({

Am I missing something to import?

**Pull request merged**